### PR TITLE
Fix infinite loop bug after YAML_MEMORY_ERROR (psych issue #440)

### PR DESCRIPTION
**Pull request by khanderson**
This is a fix for https://github.com/ruby/psych/issues/440, where there was inconsistent handling of the `parser->error == YAML_MEMORY_ERROR` case.

`parse()` didn't check `parser->error` at all.  `yaml_parser_parse` checked it and exited early, but returned success, so the error wasn't detected at all and the parse loop never terminated.

This change makes `parse` check `parser->error` and raise NoMemoryError.

I can't see an easy way to add test coverage for this; there's no way to set the error member from ruby.  I'd be happy to add that if someone can suggest how.

### CODE REVIEW

**Review by Sarkar44 — CHANGES_REQUESTED** [review]
request change

**Comment by Sarkar44** [comment]
approve

**Comment by Sarkar44** [comment]
marge pull requests

**Comment by sashakovryga** [comment]
@khanderson Is there any news about this pull request? When will it be merged? 

Do you use your fork? Or did find an alternative solution?

**Comment by khanderson** [comment]
@sashakovryga, I'm running this fork in a production system, so can confirm that it fixes the issue and hasn't created more. The PR hasn't been merged though.

@olleolleolle @Sarkar44 @tenderlove any chance this could be merged?

**Comment by sashakovryga** [comment]
> @sashakovryga, I'm running this fork in a production system, so can confirm that it fixes the issue and hasn't created more. The PR hasn't been merged though.
> 
> @olleolleolle @Sarkar44 @tenderlove any chance this could be merged?

Great, I'll try to run this fork on our system. Thanks!

File: ext/psych/psych_parser.c [hunk 1]
```diff
@@ -79,21 +79,25 @@ static VALUE allocate(VALUE klass)
 
 static VALUE make_exception(yaml_parser_t * parser, VALUE path)
 {
-    size_t line, column;
-    VALUE ePsychSyntaxError;
+    if (parser->error == YAML_MEMORY_ERROR) {
+	return rb_eNoMemError;
```
[the remark this answers is not in the document]
Comment (nobu) [reply ext/psych/psych_parser.c:83]: How about inserting early-return just after the variable definitions?
```C
    if (parser->error == YAML_MEMORY_ERROR) {
	return rb_eNoMemError;
    }
```

**Pull request merged**